src/redisearch_rs/inverted_index/src/index_result.rs

```diff
@@ -98,28 +92,66 @@ impl RSOffsetVector<'_> {
             _phantom: PhantomData,
         }
     }
+
+    /// Free the data inside this offset
+    pub fn free_data(&mut self) {
+        if self.data.is_null() {
+            return;
+        }
+
+        let layout = Layout::array::<c_char>(self.len as usize).unwrap();
+        // SAFETY: Caller is to ensure data has been allocated via the global allocator
+        // and points to an array matching the length of `offsets`.
+        unsafe { std::alloc::dealloc(self.data.cast(), layout) };
+
+        self.data = std::ptr::null_mut();
+        self.len = 0;
+    }
 }
 
 /// Represents a single record of a document inside a term in the inverted index
 /// cbindgen:rename-all=CamelCase
-#[repr(C)]
+#[repr(u8)]
 #[derive(Eq, PartialEq)]
-pub struct RSTermRecord<'index> {
-    /// We mark copied terms so we can treat them a bit differently on drop.
-    pub is_copy: bool,
+pub enum RSTermRecord<'index> {
+    Borrowed {
+        /// The term that brought up this record
+        term: *mut RSQueryTerm,
 
-    /// The term that brought up this record
-    pub term: *mut RSQueryTerm,
+        /// The encoded offsets in which the term appeared in the document
+        offsets: RSOffsetVector<'index>,
+    },
+    Owned {
+        /// The term that brought up this record
+        term: *mut RSQueryTerm,
 
-    /// The encoded offsets in which the term appeared in the document
-    pub offsets: RSOffsetVector<'index>,
+        /// The encoded offsets in which the term appeared in the document
+        offsets: RSOffsetVector<'static>,
+    },
+}
+
+impl Drop for RSTermRecord<'_> {
+    fn drop(&mut self) {
+        match self {
+            RSTermRecord::Borrowed { term, .. } => {
+                if !term.is_null() {
+                    // SAFETY: we know the C type is the same because the Rust type is autogenerated from it
+                    unsafe {
+                        Term_Free(*term);
```

Comment:
I keep being confused why the term is free in the borrowed case. May be worth properly documenting ownership in the enum definition.

Reply:
Done in https://github.com/RediSearch/RediSearch/pull/6682/commits/a6bd6f8ed25c165d8bdc242181b9cb23b739a6f0

Reply:
What's confusing me is why are we calling `Term_Free()` in the borrowed case? We should document exactly what is being owned there.

Reply:
I expanded the comments more in https://github.com/RediSearch/RediSearch/pull/6682/commits/73aa705d12d1ab9f5dfe1a4c10972de7d832576f

This is starting to become clear that `Borrowed` actually only means borrowing the offsets (even for the aggregate) and that everything else is owned. I'm thinking we should be able to find a better enum variant name here. Maybe `BorrowedOffsets`?